src/store.js

```diff
@@ -47,6 +48,7 @@ export default function initializeStore(preloadedState = undefined) {
       processingNotification: processingNotificationReducer,
       helpUrls: helpUrlsReducer,
       courseExport: courseExportReducer,
+      courseOptimizer: courseOptimizerReducer,
```

Comment:
Is there a reason for putting this in the Redux state and for implementing all this new code as JavaScript instead of TypeScript? I would really prefer that all net-new modules be using TypeScript, React Context, and React Query instead of JS and Redux, as [recommended in OEP-67](https://docs.openedx.org/projects/openedx-proposals/en/latest/best-practices/oep-0067-bp-tools-and-technology.html#frontend-technology-selection).

Is there something more I can do to raise awareness of this?

Reply:
Hi Braden, thank you so much for the feedback. I have thought this over. Let me explain where we are at.
I understand very well the wish to move us towards using TypeScript for new features. I completely agree, wholeheartedly. I was part of the fedx discussions around enabling more TypeScript.

We have very limited capacity and are looking at pushing this out fairly quickly.
We managed to base this largely on the "Export" feature that allowed us to quickly copy over code and adjust. That feature was not written in TypeScript in the first place.
I didn't succeed in using TypeScript partially for API handling here and it was taking too long to figure out.
We don't have experience with converting a lot of this functionality to TypeScript, even though I'm fluent at TypeScript. This would be a bit of a trailblazing effort but it would need quite an overhaul from the original code of the export we're basing this on.

When we were discussing whether or not to build this feature, we were working off of the promise to base this off of existing code of the export page. Needing to rewrite that in TypeScript was not something we considered and is changing the scope of this feature significantly. We also weren't aware that this would be an acceptance criteria to get a PR approved.

I'm willing to give it another try to change some very "core" functionality to TypeScript, around the API calls, but I think I would need some help with that if I need to do that? Would you be interested in doing an hour of pair programming on it sometime?

Otherwise I hope it's okay if we proceed as planned. I do think this is an important topic to factor in next time we propose a new feature. Maybe when we discuss the next feature proposal with the community we can point out the use of TypeScript from the get-go? So we can factor it in in our determination of cost vs. value.

Does that sound okay? What do you think?

Reply:
@jesperhodge 

> We also weren't aware that this would be an acceptance criteria to get a PR approved.

To be clear, I'm not going to try to block this PR until it's converted to TypeScript. It's fine as is, but in a perfect world it would have been written in TypeScript etc. in the first place.

> We managed to base this largely on the "Export" feature that allowed us to quickly copy over code and adjust. That feature was not written in TypeScript in the first place.

I see; that does make sense. If you wrote it from scratch, it would not make sense to me, because IMHO it's easier to write new things using the new patterns than the existing patterns, though obviously there is a bit of a learning curve.

What I would ask is that next time you write any REST API related code in this repo, please use one of [Library API](https://github.com/openedx/frontend-app-authoring/tree/master/src/library-authoring/data), [Taxonomy API](https://github.com/openedx/frontend-app-authoring/tree/master/src/taxonomy/data), or [Search API](https://github.com/openedx/frontend-app-authoring/tree/master/src/search-manager/data) as a starting point, as those are already written using React Query and TypeScript.

And likewise for the actual page content and tests, this repo contains a huge mix of code styles, so if you can try to pick some of the newer ones (currently, any of the Libraries, Taxonomy, or Search UI) and any tests that are written using [`testUtils`/`initializeMocks`](https://github.com/openedx/frontend-app-authoring/blob/master/src/testUtils.tsx#L152-L159), you'll be following the current best practices without any additional work.

> I'm willing to give it another try to change some very "core" functionality to TypeScript, around the API calls, but I think I would need some help with that if I need to do that? Would you be interested in doing an hour of pair programming on it sometime?

Again, not a requirement, but - that would be great! See the examples I mentioned above and ping me on Slack anytime you want to do pair programming or have any questions I can answer.

Reply:
Also, you should know that you can put `// @ts-check` as the first line of your `.jsx` files, and you'll still get many of the benefits of using TypeScript in that file (e.g. code completion, type checking, and inline documentation in your IDE) without having to do the work of converting the file to TypeScript. It does make things stricter though, so sometimes you need to use JSDoc annotations and type casting to eliminate the warnings after making that change.